### ISSUE TITLE
git versioning tags not explained

### ISSUE THREAD

**Issue by yacc143** [python-poetry/poetry#3403]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.

## Issue

Can somebody explain the difference between:

`uberdbutils (0.0.22 git branch master)`

and

`uberdbutils (0.0.22 git rev master)`

Especially as it produces a conflict that makes it impossible to upgrade a package:

`  Because uber-schema (2.1.1) depends on uberdbutils (0.0.22 git rev master)
   and no versions of uber-schema match >2.1.1,<3.0.0, uber-schema (>=2.1.1,<3.0.0) requires uberdbutils (0.0.22 git rev master).
  So, because grants-pipeline depends on both uberdbutils (0.0.22 git branch master) and uber-schema (^2.1.1), version solving failed.
`
The package blocks for uberdbutils btw in poetry.lock in grants-pipeline & uber-schema are exactly the same.

So this is driving my crazy, uberdbutils is referenced by uber-schema & grants-pipeline in pyproject.toml & poetry.lock exactly the same, but poetry calls it 0.0.22 git rev master here and 0.0.22 git branch master there???

(And no, there have been no commits to uberdbutils has not been updated in 18 months.

**Comment by rambo** [comment]
I have a related problem with git version resolving failing:

"""Because PrivatePackageOne (0.3.2 git rev 0.3.2) depends on PrivatePackageTwo (0.3.1 git tag 0.3.1)
 and PrivateProject depends on PrivatePackageTwo (0.3.1 git tag 0.3.1), PrivatePackageOne is forbidden.
So, because PrivateProject depends on PrivatePackageOne (0.3.2 git tag 0.3.2), version solving failed."""

Where PrivateProject is the current project/package name I'm trying to run `poetry update` on, PrivatePackageOne and PrivatePackageTwo are git deps with `tag="0.3.x"` (not just branch but I have seen the same problem with deps pointing to the same branch) both of those are also package with poetry.

In this case is PrivatePackageOne devel dependency and PrivatePackageTwo is normal dep so I can't just remove PrivatePackageTwo from deps completely.

**Comment by sinoroc** [comment]
@rambo probably related: https://github.com/python-poetry/poetry/pull/3389

**Comment by yacc143** [comment]
Sadly, in my case I've used exactly the same form in both cases. But I'll
try to specify an explicit branch=master to see if that helps.

Eero af Heurlin <notifications@github.com> schrieb am Mo., 23. Nov. 2020,
17:01:

**Comment by sinoroc** [comment]
I see this line of code, and I wonder if there is any difference in the handling at all:

```
source_reference=branch or tag or rev,
```
-- https://github.com/python-poetry/poetry-core/blob/c7a4c31684f00e7f3a677d922b73fbead91528d2/poetry/core/packages/vcs_dependency.py#L49

But I do not know enough about _poetry_'s code to make an informed judgement here.

I also wonder if git itself handles those differently for the checkout and/or whatever git commands are used in such cases.

**Comment by yacc143** [comment]
Ok, I think I've understood the difference between rev/branch (although I was not able to change it both to branch, I was able to change it both to "rev", and now the resolver error message makes simply no sense:

`  Because uber-schema (2.1.4) depends on uberdbutils (0.0.22 git rev master)
   and no versions of uber-schema match >2.1.4,<3.0.0, uber-schema (>=2.1.4,<3.0.0) requires uberdbutils (0.0.22 git rev master).
  So, because grants-pipeline depends on both uberdbutils (0.0.22 git rev master) and uber-schema (^2.1.4), version solving failed.
`

To summarize in simpler words because B depends on C and no versions of B matching require C, A requires B & C, so version solving failed????????

**Comment by yacc143** [comment]
Btw, please note, poetry had no problem to find the initial version of the packages involved, it just refuses to upgrade them because of the common git repo dependency. Weep.
